File: ts/project/.sdk/tm/scala/feature/FeatureOptions.scala
package SCALAPACKAGE.feature

import java.util.{ArrayList, List => JList, Map => JMap}
import java.util.function.{IntConsumer, LongSupplier}

// Shared option readers for the feature implementations.
object FeatureOptions {

  def foptBool(options: JMap[String, Object], key: String, dflt: Boolean): Boolean = {
    if (options == null) return dflt
    options.get(key) match { case b: java.lang.Boolean => b.booleanValue(); case _ => dflt }
  }

  def foptInt(options: JMap[String, Object], key: String, dflt: Int): Int = {
    if (options == null) return dflt
    options.get(key) match { case n: java.lang.Number => n.intValue(); case _ => dflt }
  }

  def foptNum(options: JMap[String, Object], key: String, dflt: Double): Double = {
    if (options == null) return dflt
    options.get(key) match { case n: java.lang.Number => n.doubleValue(); case _ => dflt }
  }

  def foptStr(options: JMap[String, Object], key: String, dflt: String): String = {
    if (options == null) return dflt
    options.get(key) match { case s: String if s != "" => s; case _ => dflt }
  }

  def foptMap(options: JMap[String, Object], key: String): JMap[String, Object] = {
    if (options == null) return null
    options.get(key) match { case m: JMap[_, _] => m.asInstanceOf[JMap[String, Object]]; case _ => null }
  }

  def foptList(options: JMap[String, Object], key: String): JList[Object] = {
    if (options == null) return null
    options.get(key) match { case l: JList[_] => l.asInstanceOf[JList[Object]]; case _ => null }
  }

  def foptStrList(options: JMap[String, Object], key: String): JList[String] = {
    val raw = foptList(options, key)
    if (raw == null) return null
    val out = new ArrayList[String]()
    val it = raw.iterator()
    while (it.hasNext) it.next() match { case s: String => out.add(s); case _ => }
    out
  }

  // Injectable sleep (option "sleep": IntConsumer of ms), defaulting to a
  // real Thread.sleep.
  def foptSleep(options: JMap[String, Object]): IntConsumer = {
    if (options != null) options.get("sleep") match { case ic: IntConsumer => return ic; case _ => }
    (ms: Int) => if (ms > 0) {
      try Thread.sleep(ms.toLong)
      catch { case _: InterruptedException => Thread.currentThread().interrupt() }
    }
  }

  // Injectable clock (option "now": LongSupplier of ms), defaulting to the
  // wall clock.
  def foptNow(options: JMap[String, Object]): LongSupplier = {
    if (options != null) options.get("now") match { case ls: LongSupplier => return ls; case _ => }
    () => System.currentTimeMillis()
  }

  def fheaderGet(headers: JMap[String, Object], name: String): Object = {
    if (headers == null) return null
    val it = headers.entrySet().iterator()
    while (it.hasNext) {
      val e = it.next()
      if (e.getKey != null && e.getKey.equalsIgnoreCase(name)) return e.getValue
    }
    null
  }

  def fheaderHas(headers: JMap[String, Object], name: String): Boolean = {
    if (headers == null) return false
    val it = headers.keySet().iterator()
    while (it.hasNext) {
      val k = it.next()
      if (k != null && k.equalsIgnoreCase(name)) return true
    }
    false
  }

  def fheaderSetDefault(headers: JMap[String, Object], name: String, value: String): Unit = {
    if (headers == null) return
    if (fheaderHas(headers, name)) return
    headers.put(name, value)
  }

  def fresStatus(res: Object): Int = res match {
    case m: JMap[_, _] =>
      m.asInstanceOf[JMap[String, Object]].get("status") match { case n: java.lang.Number => n.intValue(); case _ => -1 }
    case _ => -1
  }

  def fresHeader(res: Object, name: String): String = res match {
    case m: JMap[_, _] =>
      m.asInstanceOf[JMap[String, Object]].get("headers") match {
        case hm: JMap[_, _] =>
          fheaderGet(hm.asInstanceOf[JMap[String, Object]], name) match { case s: String => s; case _ => "" }
        case _ => ""
      }
    case _ => ""
  }

  def fparseInt(s: String, dflt: Int): Int = {
    try Integer.parseInt(s.trim)
    catch { case _: RuntimeException => dflt }
  }
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/FeatureOptions.swift
// ProjectName SDK - shared option readers for the feature implementations.
// Feature options arrive as a loose Value map; these helpers normalise access
// and supply defaults, mirroring the C# donor's FeatureOptions.

import Foundation

func fopt(_ options: VMap?, _ key: String) -> Value { gp(options, key) }

func foptBool(_ options: VMap?, _ key: String, _ def: Bool) -> Bool {
  gp(options, key).asBool ?? def
}

func foptInt(_ options: VMap?, _ key: String, _ def: Int) -> Int {
  switch gp(options, key) {
  case .int(let n): return Int(n)
  case .double(let d): return Int(d)
  default: return def
  }
}

func foptNum(_ options: VMap?, _ key: String, _ def: Double) -> Double {
  switch gp(options, key) {
  case .int(let n): return Double(n)
  case .double(let d): return d
  default: return def
  }
}

func foptStr(_ options: VMap?, _ key: String, _ def: String) -> String {
  if let s = gp(options, key).asString, s != "" { return s }
  return def
}

func foptMap(_ options: VMap?, _ key: String) -> VMap? { gp(options, key).asMap }

func foptList(_ options: VMap?, _ key: String) -> VList? { gp(options, key).asList }

func foptStrList(_ options: VMap?, _ key: String) -> [String]? {
  guard let l = gp(options, key).asList else { return nil }
  return l.items.compactMap { $0.asString }
}

// FoptSleep returns the injectable sleep (option "sleep": (Int) -> Void),
// defaulting to a real Thread.sleep.
func foptSleep(_ options: VMap?) -> (Int) -> Void {
  if let fn = gp(options, "sleep").asNative as? (Int) -> Void { return fn }
  return { ms in if ms > 0 { Thread.sleep(forTimeInterval: Double(ms) / 1000.0) } }
}

// FoptNow returns the injectable clock (option "now": () -> Int64, ms),
// defaulting to the wall clock.
func foptNow(_ options: VMap?) -> () -> Int64 {
  if let fn = gp(options, "now").asNative as? () -> Int64 { return fn }
  return { Int64(Date().timeIntervalSince1970 * 1000) }
}

// FheaderGet reads a header value case-insensitively.
func fheaderGet(_ headers: VMap?, _ name: String) -> (Value, Bool) {
  guard let headers = headers else { return (.noval, false) }
  for (k, v) in headers.entries {
    if k.caseInsensitiveCompare(name) == .orderedSame { return (v, true) }
  }
  return (.noval, false)
}

// FheaderSetDefault sets a header only when no case-insensitive variant exists.
func fheaderSetDefault(_ headers: VMap?, _ name: String, _ value: String) {
  guard let headers = headers else { return }
  let (_, has) = fheaderGet(headers, name)
  if has { return }
  headers.entries[name] = .string(value)
}

// FresStatus extracts the numeric status from a transport-shaped response.
func fresStatus(_ res: Value) -> (Int, Bool) {
  guard let rm = res.asMap, let raw = rm.entries["status"] else { return (0, false) }
  switch raw {
  case .int(let n): return (Int(n), true)
  case .double(let d): return (Int(d), true)
  default: return (0, false)
  }
}

// FresHeader reads a header from a transport-shaped response.
func fresHeader(_ res: Value, _ name: String) -> (String, Bool) {
  guard let rm = res.asMap, let headers = rm.entries["headers"]?.asMap else { return ("", false) }
  let (v, has) = fheaderGet(headers, name)
  if !has { return ("", false) }
  guard let str = v.asString else { return ("", false) }
  return (str, true)
}

// FparseInt parses a decimal string; def when unparseable.
func fparseInt(_ s: String, _ def: Int) -> Int {
  Int(s.trimmingCharacters(in: .whitespaces)) ?? def
}
